### ISSUE TITLE
libgomp problem

### ISSUE THREAD

**Issue by murenrb** [CoolProp/CoolProp#1139]
Hey Coolprop team. 
I am not sure if this is a coolprop issue, but we are running into a problem via coolprop. After a certain number of calls we get the error:

libgomp: could not create thread pool destructor.

We are using python via pycharm on OSX ElCap. The strange thing is that this error happens after a certain number of calculations. IE, if we make our code converge fast enough, than we will not get the error. But if we, for example, put in only mediocre guess values the error will occur after a pretty predictable time. 

The error seems to be happening in the coolprop call PropSI we are using to find saturation conditions of a fluid via refprop data using librefprop. We posted a [stack overflow](http://stackoverflow.com/questions/33962664/python-error-unknown-thread-pool-destructor) question about this and it was a total bust. So hopefully someone has some ideas? It is making all our code pretty useless.

**Comment by ibell** [comment]
Hmm I've never seen this before.  Sounds like something that is not a problem arising from CoolProp.  We don't use any multi-threading at all in the CoolProp core.

In the past I have done some multi-processing stuff with CoolProp, and that has been successful on linux/windows/OSX.  Could you do your calculations with something like scoop/multiprocessing instead of of threading?

Are you calling REFPROP through CoolProp? REFPROP does use OMP.

**Comment by murenrb** [comment]
Sorry for not being clearer before, we are calling REFPROP via CoolProp. We are using LibRefProp to do that. We don't use any threading in our own code, so that is why our best guess was CoolProp because it seems to happen when the code is in a PropSI call. We only know this from print statements as there is no traceback.

Do you think this is an issue in Librefprop? or REFPROP itself? 

If there is anything we can send you to help you figure it out let us know because it is very opaque for us. We know it is happening in the PropSI command and it happens on two of our computers. Or perhaps we should email the Refprop folks? I don't have a good idea of where the CoolProp / Refprop boundary is. 

Thanks for your help!!

**Comment by ibell** [comment]
Sounds like a REFPROP problem, not a CoolProp problem.  Could you perhaps build REFPROP with this cmake setup: https://github.com/usnistgov/REFPROP-cmake .  

In my day job I work at NIST (the REFPROP folks).  Do you have a small file that you could give me that would demonstrate this failure?  Otherwise its impossible to debug.

**Comment by JonWel** [comment]
@murenrb  Not related to your issue, but for iterative solver using the [low level interface](http://www.coolprop.org/coolprop/LowLevelAPI.html) has speed advantages against PropsSI.
PropsSI has to initialise the fluid at each iteration, whereas with the low level API you can initialise once before the loop. You gain even more if you need several values at each iterations as you can have only one state update (PropsSI is doing all operations at each call).


On the problem's side, I saw this on your [stackoverflow](http://stackoverflow.com/questions/33962664/python-error-unknown-thread-pool-destructor) thread: 

> I have run the same code on 2 other macs running the same operating system and the same setups. I setup all 3 the same. The other two have no issue.

If you cannot replicate on other computers, it may also be difficult for REFPROP's guy to replicate.


I'm also seeing this in the answer you get:

> only called (once!) at process/libgomp initialization

@ibell could REFPROP's PropsSI call initialise this `pthread_key_create` at each call? If this is the case, using the low level API could be a workaround (as I expect it would call only once at abstractclass's initialisation).

**Comment by ibell** [comment]
@JonWel has a pretty good idea.  It's always best to use the low-level
interface when possible. It's more computationally efficient (most
especially for REFPROP).

**Comment by murenrb** [comment]
Here is a quick code to duplicate the issue:

```
from CoolProp.CoolProp import PropsSI
from random import random

fluid = 'REFPROP-MIX:AMMONIA&WATER'
P=100000.0 #Pa
c=random()
mw1 = float(PropsSI('M', 'T', 273, 'P', 101325., fluid[fluid.find('&')+1:fluid.__len__()]))
mw2 = float(PropsSI('M', 'T', 273, 'P', 101325, fluid[fluid.find(':')+1:fluid.find('&')]))
i=0
while (i < 1000):
    i+=1
    if fluid[-1] is not ']':        # add concentrations to fluid name if not already done
        if c == 0.0:
            c_mol = 0.0
        else:
            c_mol = (c/mw2)/(c/mw2 + (1.-c)/mw1)
        fluid = fluid.replace('&', '[%s]&'%str(c_mol).strip()) + '[%s]'%str(1.0-c_mol).strip()

    tlsat = PropsSI('T', 'Q', 0.0, 'P', P, fluid)-273.15
    if isinstance(i/10, (int, long)):
        print c_mol,tlsat, i
```

We have looked into the low level interface before, but unfortunately we are using a library that required `PropSI` calls. We could probably re-write the structure to use the low-level interface, but it is a very large undertaking and we have fairly extreme bandwidth issues. 

At any rate, we are now seeing this issue on 3 of our 5 engineer's systems, so it appears to be more wide spread than originally noted in the first stackoverflow question. 

When I run the above code, I get the error on run 510 every time. It seems to be very repeatable. 

All this said, it appears that using CMake solves the issue. @ericksonla was able to get rid of the issue by switching. So perhaps this is just another one of those mysteries that will fade into the unknowable. At any rate, it does not appear to have been an issue with CoolProp so this issue can be closed. 

as always thanks for all your help!

**Comment by ibell** [comment]
What version of REFPROP do you use?  I know there was a memory leak in REFPROP some time ago with calling SETUPdll function too often, but newer versions fixed it.  Perhaps that is in some fashion responsible for this failure.  In any case, the low-level interface should call SETUPdll much less often if you use it properly.

Glad to hear the REFPROP-cmake build "fixes" this issue.  Perhaps it was just a question of the age of the REFPROP compiler you were using.
